File: benchmarks/cross-language/erlang/skynet.erl
-module(skynet).
-export([start/0, skynet/4]).

-define(BASE, 10).

skynet(Level, Num, Parent, Count) ->
    if
        Level >= 4 ->
            Parent ! {result, Num + Count};
        true ->
            NewLevel = Level + 1,
            spawn_children(NewLevel, Num, self(), Count, ?BASE, 0)
    end.

spawn_children(_, _, Parent, _Count, 0, Sum) ->
    Parent ! {result, Sum};
spawn_children(Level, Num, Parent, Count, Remaining, Sum) ->
    Child = Remaining - 1,
    spawn(?MODULE, skynet, [Level, Num * ?BASE + Child, Parent, Count]),
    receive
        {result, ChildSum} ->
            spawn_children(Level, Num, Parent, Count, Child, Sum + ChildSum)
    end.

collect_results(0, Sum) ->
    Sum;
collect_results(Remaining, Sum) ->
    receive
        {result, ChildSum} ->
            collect_results(Remaining - 1, Sum + ChildSum)
    end.

start() ->
    io:format("=== Erlang Skynet Benchmark ===~n"),
    io:format("Base: ~p~n", [?BASE]),
    io:format("Depth: 4~n~n", []),

    Count = 1000,
    ExpectedActors = 1 + ?BASE + (?BASE * ?BASE) + (?BASE * ?BASE * ?BASE),
    TotalMessages = ExpectedActors * Count,

    io:format("Expected actors: ~p~n", [ExpectedActors]),
    io:format("Messages per actor: ~p~n", [Count]),
    io:format("Total messages: ~p~n~n", [TotalMessages]),

    % Start timer
    Start = erlang:system_time(nanosecond),

    % Spawn root actors
    Parent = self(),
    lists:foreach(fun(I) ->
        spawn(?MODULE, skynet, [0, I, Parent, Count])
    end, lists:seq(0, ?BASE - 1)),

    % Collect results
    _Sum = collect_results(?BASE, 0),

    % Calculate metrics
    End = erlang:system_time(nanosecond),
    ElapsedNs = End - Start,
    ElapsedSec = ElapsedNs / 1000000000.0,

    % Estimate cycles (assuming 3GHz)
    TotalCycles = ElapsedNs * 3.0,
    CyclesPerMsg = TotalCycles / TotalMessages,
    MsgPerSec = TotalMessages / ElapsedSec,

    io:format("Cycles/msg:     ~.2f~n", [CyclesPerMsg]),
    io:format("Throughput:     ~.0f M msg/sec~n", [MsgPerSec / 1000000]),

    halt(0).

File: benchmarks/cross-language/rust/src/skynet.rs
use tokio::sync::mpsc;
use tokio::task;
use std::time::Instant;

const BASE: usize = 10;
const COUNT: i64 = 1000;

async fn skynet(level: usize, num: usize, tx: mpsc::Sender<i64>) {
    if level >= 4 {
        let _ = tx.send(num as i64 + COUNT).await;
    } else {
        let (result_tx, mut result_rx) = mpsc::channel::<i64>(BASE);

        for i in 0..BASE {
            let child_tx = result_tx.clone();
            task::spawn(skynet(level + 1, num * BASE + i, child_tx));
        }
        drop(result_tx);

        let mut sum = 0;
        while let Some(val) = result_rx.recv().await {
            sum += val;
        }

        let _ = tx.send(sum).await;
    }
}

#[tokio::main]
async fn main() {
    println!("=== Rust Skynet Benchmark ===");
    println!("Base: {}", BASE);
    println!("Depth: 4");

    let expected_actors = 1 + BASE + (BASE * BASE) + (BASE * BASE * BASE);
    let total_messages = expected_actors * COUNT as usize;

    println!("\nExpected actors: {}", expected_actors);
    println!("Messages per actor: {}", COUNT);
    println!("Total messages: {}\n", total_messages);

    let (tx, mut rx) = mpsc::channel::<i64>(BASE);

    let start = Instant::now();

    for i in 0..BASE {
        let child_tx = tx.clone();
        task::spawn(skynet(0, i, child_tx));
    }
    drop(tx);

    let mut _sum = 0;
    while let Some(val) = rx.recv().await {
        _sum += val;
    }

    let elapsed = start.elapsed();

    let cycles = (elapsed.as_secs_f64() * 3e9) as u64;
    let cycles_per_msg = cycles as f64 / total_messages as f64;
    let msg_per_sec = total_messages as f64 / elapsed.as_secs_f64();

    println!("Cycles/msg:     {:.2}", cycles_per_msg);
    println!("Throughput:     {:.0} M msg/sec", msg_per_sec / 1_000_000.0);
}
